### PR TITLE
fix(backends): fix null literal handling

### LINKED ISSUE #5520: bug: `ibis.literal(None, type="float64")` errors
### What happened?

I get
```python backtrace
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/IPython/core/formatters.py:706, in PlainTextFormatter.__call__(self, obj)
    699 stream = StringIO()
    700 printer = pretty.RepresentationPrinter(stream, self.verbose,
    701     self.max_width, self.newline,
    702     max_seq_length=self.max_seq_length,
    703     singleton_pprinters=self.singleton_printers,
    704     type_pprinters=self.type_printers,
    705     deferred_pprinters=self.deferred_printers)
--> 706 printer.pretty(obj)
    707 printer.flush()
    708 return stream.getvalue()

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/IPython/lib/pretty.py:410, in RepresentationPrinter.pretty(self, obj)
    407                         return meth(obj, self, cycle)
    408                 if cls is not object \
    409                         and callable(cls.__dict__.get('__repr__')):
--> 410                     return _repr_pprint(obj, self, cycle)
    412     return _default_pprint(obj, self, cycle)
    413 finally:

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/IPython/lib/pretty.py:778, in _repr_pprint(obj, p, cycle)
    776 """A pprint that just redirects to the normal repr function."""
    777 # Find newlines and replace them with p.break_()
--> 778 output = repr(obj)
    779 lines = output.splitlines()
    780 with p.group():

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/expr/types/core.py:54, in Expr.__repr__(self)
     52 with simple_console.capture() as capture:
     53     try:
---> 54         simple_console.print(self)
     55     except TranslationError as e:
     56         lines = [
     57             "Translation to backend failed",
     58             f"Error message: {e.args[0]}",
     59             "Expression repr follows:",
     60             self._repr(),
     61         ]

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/rich/console.py:1694, in Console.print(self, sep, end, style, justify, overflow, no_wrap, emoji, markup, highlight, width, height, crop, soft_wrap, new_line_start, *objects)
   1692 if style is None:
   1693     for renderable in renderables:
-> 1694         extend(render(renderable, render_options))
   1695 else:
   1696     for renderable in renderables:

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/rich/console.py:1306, in Console.render(self, renderable, options)
   1304 renderable = rich_cast(renderable)
   1305 if hasattr(renderable, "__rich_console__") and not isclass(renderable):
-> 1306     render_iterable = renderable.__rich_console__(self, _options)  # type: ignore[union-attr]
   1307 elif isinstance(renderable, str):
   1308     text_renderable = self.render_str(
   1309         renderable, highlight=_options.highlight, markup=_options.markup
   1310     )

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/expr/types/generic.py:562, in Scalar.__rich_console__(self, console, options)
    560 if not ibis.options.interactive:
    561     return console.render(Text(self._repr()), options=options)
--> 562 return console.render(repr(self.execute()), options=options)

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/expr/types/core.py:300, in Expr.execute(self, limit, timecontext, params, **kwargs)
    273 def execute(
    274     self,
    275     limit: int | str | None = 'default',
   (...)
    278     **kwargs: Any,
    279 ):
    280     """Execute an expression against its backend if one exists.
    281 
    282     Parameters
   (...)
    298         Keyword arguments
    299     """
--> 300     return self._find_backend(use_default=True).execute(
    301         self, limit=limit, timecontext=timecontext, params=params, **kwargs
    302     )

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/backends/base/sql/__init__.py:234, in BaseSQLBackend.execute(self, expr, params, limit, **kwargs)
    232 kwargs.pop('timecontext', None)
    233 query_ast = self.compiler.to_ast_ensure_limit(expr, limit, params=params)
--> 234 sql = query_ast.compile()
    235 self._log(sql)
    237 schema = self.ast_schema(query_ast, **kwargs)

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/backends/base/sql/compiler/base.py:40, in QueryAST.compile(self)
     38 def compile(self):
     39     compiled_setup_queries = [q.compile() for q in self.setup_queries]
---> 40     compiled_queries = [q.compile() for q in self.queries]
     41     compiled_teardown_queries = [q.compile() for q in self.teardown_queries]
     42     return self.context.collapse(
     43         list(
     44             chain(
   (...)
     49         )
     50     )

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/backends/base/sql/compiler/base.py:40, in <listcomp>(.0)
     38 def compile(self):
     39     compiled_setup_queries = [q.compile() for q in self.setup_queries]
---> 40     compiled_queries = [q.compile() for q in self.queries]
     41     compiled_teardown_queries = [q.compile() for q in self.teardown_queries]
     42     return self.context.collapse(
     43         list(
     44             chain(
   (...)
     49         )
     50     )

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/backends/base/sql/alchemy/query_builder.py:185, in AlchemySelect.compile(self)
    176 steps = [
    177     self._add_select,
    178     self._add_group_by,
   (...)
    181     self._add_limit,
    182 ]
    184 for step in steps:
--> 185     frag = step(frag)
    187 return frag

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/backends/base/sql/alchemy/query_builder.py:218, in AlchemySelect._add_select(self, table_set)
    216 for op in select_set:
    217     if isinstance(op, ops.Value):
--> 218         arg = self._translate(op, named=True)
    219     elif isinstance(op, ops.TableNode):
    220         arg = context.get_ref(op)

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/backends/base/sql/compiler/query_builder.py:218, in Select._translate(self, expr, named, permit_subquery)
    211 def _translate(self, expr, named=False, permit_subquery=False):
    212     translator = self.translator_class(
    213         expr,
    214         context=self.context,
    215         named=named,
    216         permit_subquery=permit_subquery,
    217     )
--> 218     return translator.get_result()

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/backends/base/sql/compiler/translator.py:211, in ExprTranslator.get_result(self)
    209 def get_result(self):
    210     """Compile SQL expression into a string."""
--> 211     translated = self.translate(self.node)
    212     if self._needs_name(self.node):
    213         # TODO: this could fail in various ways
    214         translated = self.name(translated, self.node.name)

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/backends/base/sql/compiler/translator.py:243, in ExprTranslator.translate(self, op)
    241 elif type(op) in self._registry:
    242     formatter = self._registry[type(op)]
--> 243     return formatter(self, op)
    244 else:
    245     raise com.OperationNotDefinedError(f'No translation rule for {type(op)}')

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/backends/duckdb/registry.py:107, in _literal(t, op)
    105     return sa.cast(sa.func.list_value(*values), sqla_type)
    106 elif dtype.is_floating():
--> 107     if not np.isfinite(value):
    108         if np.isnan(value):
    109             value = "NaN"

TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

I have to use  `ibis.null()` instead of None, which isn't a huge deal, but the error message is super cryptic so I barely figured out that this was the root cause.

From looking at `def _literal(t, op)` in `ibis/backends/duckdb/registry.py`, it looks like structs and arrays also are gonna have this problem, and from testing they do.

### What version of ibis are you using?

main

### What backend(s) are you using, if any?

duckdb, this one seems specific to duckdb. IDK about the other backends they might be susceptible too, I didn;t look

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by cpcloud**
Fix backend implementations of NULL literal handling. Closes #5520.

### CODE REVIEW

**Comment by cpcloud** [comment]
Ok, I'm going to leave bool handling here for a follow-up, because it's broken in a deeper way than should be fixed here. It won't really be fixed until we can use pandas nullable dtypes.

File: ibis/backends/bigquery/registry.py [hunk 1]
```diff
@@ -221,26 +221,30 @@ def _log(translator, op):
...
-            return f"{ibis_type_to_bigquery_type(dtype)} '{op.value}'"
+            return f"{ibis_type_to_bigquery_type(dtype)} '{value}'"
     elif dtype.is_uuid():
-        return translator.translate(ops.Literal(str(op.value), dtype=dt.str))
+        return translator.translate(ops.Literal(str(value), dtype=dt.str))
 
     if isinstance(dtype, dt.Numeric):
-        value = op.value
+        if value is None:
+            return value
```
Comment (NickCrews) [inline ibis/backends/bigquery/registry.py:242]: Isn't this case already covered?

Comment (cpcloud) [reply ibis/backends/bigquery/registry.py:242]: Yep, thanks!

**Comment by NickCrews** [comment]
Looks great, thank you @cpcloud!

**Review by kszucs — APPROVED** [review]
Nice!

**Pull request merged**